src/Storages/PostgreSQL/PostgreSQLReplicaConsumer.cpp

```diff
@@ -0,0 +1,682 @@
+                insertDefaultValue(buffer, column_idx);
+                break;
+            }
+            case 't': /// Text formatted value
+            {
+                Int32 col_len = readInt32(message, pos, size);
+                String value;
+
+                for (Int32 i = 0; i < col_len; ++i)
+                    value += readInt8(message, pos, size);
+
+                insertValue(buffer, value, column_idx);
+                break;
+            }
+            case 'u': /// TOAST value && unchanged at the same time. Actual value is not sent.
+                /// TOAST values are not supported. (TOAST values are values that are considered in postgres
+                /// to be too large to be stored directly)
+                insertDefaultValue(buffer, column_idx);
+                break;
+        }
+    };
+
+    for (int column_idx = 0; column_idx < num_columns; ++column_idx)
+        proccess_column_value(readInt8(message, pos, size), column_idx);
+
+    switch (type)
+    {
+        case PostgreSQLQuery::INSERT:
+        {
+            buffer.columns[num_columns]->insert(Int8(1));
+            buffer.columns[num_columns + 1]->insert(UInt64(metadata.version()));
+
+            break;
+        }
+        case PostgreSQLQuery::DELETE:
+        {
+            buffer.columns[num_columns]->insert(Int8(-1));
+            buffer.columns[num_columns + 1]->insert(UInt64(metadata.version()));
+
+            break;
+        }
+        case PostgreSQLQuery::UPDATE:
+        {
+            /// Process old value in case changed value is a primary key.
+            if (old_value)
+                buffer.columns[num_columns]->insert(Int8(-1));
+            else
+                buffer.columns[num_columns]->insert(Int8(1));
+
+            buffer.columns[num_columns + 1]->insert(UInt64(metadata.version()));
+
+            break;
+        }
+    }
+}
+
+
+/// https://www.postgresql.org/docs/13/protocol-logicalrep-message-formats.html
+void PostgreSQLReplicaConsumer::processReplicationMessage(const char * replication_message, size_t size)
+{
+    /// Skip '\x'
+    size_t pos = 2;
+    char type = readInt8(replication_message, pos, size);
+
+    switch (type)
+    {
+        case 'B': // Begin
+        {
+            readInt64(replication_message, pos, size); /// Int64 transaction end lsn
+            readInt64(replication_message, pos, size); /// Int64 transaction commit timestamp
+            break;
+        }
+        case 'I': // Insert
+        {
+            Int32 relation_id = readInt32(replication_message, pos, size);
+
+            if (!isSyncAllowed(relation_id))
+                return;
+
+            Int8 new_tuple = readInt8(replication_message, pos, size);
+            const auto & table_name = relation_id_to_name[relation_id];
+            auto buffer = buffers.find(table_name);
+
+            if (buffer == buffers.end())
+                throw Exception(ErrorCodes::UNKNOWN_TABLE, "Buffer for table {} does not exist", table_name);
```

Comment:
When may it happen? Why don't we check if buffer exists when processing `UPDATE` and `DELETE`? Maybe replace `UNKNOWN_TABLE` with `LOGICAL_ERROR`?

Reply:
It cannot happen, forgot to remove.

Reply:
It's probably better to replace it with `assert` (or `LOGICAL_ERROR`), than just remove.

Reply:
I think it is redundant to add this check for each message because:
There might be no buffer only if there is no such storage. In this case it will be checked in message type `R`, where `relation_id` is added into skip_list if there is no such storage. This skip_list is checked for each message, so there is no need for buffer check.

Reply:
Ok, I added `assert's` for each message.